obvspython/lenses.py

```diff
@@ -0,0 +1,191 @@
+                    layer_preds.append(top_pred)
+
+                logits.append(layer_logits)
+                preds.append(layer_preds)
+
+            x_ticks = [f'{self.patchscope.tokenizer.decode(tok)}'
+                       for tok in self.data['substring_tokens']]
+            y_ticks = [f'{self.patchscope.MODEL_SOURCE}_{self.patchscope.LAYER_SOURCE}{i}'
+                       for i in self.data['layers']]
+
+            # create a heatmap with the top logits and predicted tokens
+            fig = create_annotated_heatmap(logits, preds, x_ticks, y_ticks,
+                                           title='Top predicted token and its logit')
+
+        if file_name:
+            fig.write_html(f'{file_name.replace(".html", "")}.html')
+        return fig
+
+
+class PatchscopeLogitLens(BaseLogitLens):
+    """ Implementation of logit-lens in patchscope framework.
+        The logit-lens is defined in the patchscope framework as follows:
+        S = T   (source prompt = target prompt)
+        M = M*  (source model = target model)
+        l* = L* (target layer = last layer)
+        i = i*  (source position = target position)
+        f = id  (mapping = identity function)
+
+        The source layer l and position i can vary.
+
+        In words: The logit-lens maps the hidden state at position i of layer l of the model M
+            to the last layer of that same model. It is equal to taking the hidden state and
+            applying unembed to it.
+        """
+
+    def run(self, substring: str, layers: List[int]):
+        """ Run the logit lens for each layer in layers and each token in substring.
+
+        Args:
+            substring (str): Substring of the prompt for which the top prediction and logits
+                should be calculated.
+            layers (List[int]): Indices of Transformer Layers for which the lens should be applied
+        """
+
+        # get starting position and tokens of substring
+        start_pos, substring_tokens = self.patchscope.source_position_tokens(substring)
+
+        self.data['logits'] = {}
+
+        # loop over each layer and token in substring
+        for i, layer in enumerate(layers):
+            for j in range(len(substring_tokens)):
+
+                self.patchscope.source.layer = layer
+                self.patchscope.source.position = start_pos + j
+                self.patchscope.target.position = start_pos + j
+                self.patchscope.run()
+
+                self.data['logits'][(i, j)] = self.patchscope.logits()[start_pos + j].to('cpu')
+
+            # empty CDUA cache to avoid filling of GPU memory
+            torch.cuda.empty_cache()
+        self.data['substring_tokens'] = substring_tokens
+        self.data['layers'] = layers
+
+
+class ClassicLogitLens(BaseLogitLens):
+    """ Implementation of LogitLens in standard fashion.
+        Run a forward pass on the model and multiply the output of a specific layer
```

Comment:
multiply here is wrong isnt it?

Reply:
or you just mean to apply the final layer norm?

Reply:
Yeah, I mean just apply the final layer norm. I can change it.